### PR TITLE
fix: check for invalid arguments in `validate`

### LINKED ISSUE #924: fix: add more validations rules to `validate`

### PR DESCRIPTION
**Pull request by wa0x6e**
## 🧿 Current issues / What's wrong ?

From https://github.com/orgs/snapshot-labs/projects/12/views/7?pane=issue&itemId=43073796

The `validate` passes the arguments as is to score-api without check. There are some basic check possible from the SDK, to filter out invalid requests, and lighten the score-api server load.

This will also avoid log pollution on score-api, making the log more relevant

## 💊 Fixes / Solution

Fix #924

Add some basic argument validation on the SDK side, and return a meaningful error when argument are not valid.

## 🚧 Changes

- Add check that author address is valid
- Add check that network is not empty, and valid (referenced in networks.json)
- Check that the given `snapshot` is in the network range (should be greater than the network's `start` block)
- Check that `network` in strategies are valid
- Fully tests `validate`

## 🛠️ Tests

- Send some request to `validate` functions, with invalid data
- The function should throw an error with an error message

### CODE REVIEW

**Comment by wa0x6e** [comment]
@ChaituVR do we allow empty author (0x000…) or should we also add a check for that?

**Comment by ChaituVR** [comment]
> @ChaituVR do we allow empty author (0x000…) or should we also add a check for that?

We have a check on score-api is it important to check everything here also?

**Comment by wa0x6e** [comment]
> > @ChaituVR do we allow empty author (0x000…) or should we also add a check for that?
> 
> We have a check on score-api is it important to check everything here also?

This one is less important, but does not cost to add, and benefit will be less requests to score-api

**Comment by ChaituVR** [comment]
Lets add it then 😅

File: src/utils.ts [hunk 1]
```diff
@@ -333,6 +334,19 @@ export async function validate(
...
+  if (!isValidAddress(author)) {
+    return Promise.reject(`Invalid author: ${author}`);
+  }
+
+  if (!isValidNetwork(network)) {
+    return Promise.reject(`Invalid network: ${network}`);
+  }
+  if (!isValidSnapshot(snapshot, network)) {
+    return Promise.reject(
+      `Snapshot (${snapshot}) must be greater than network start block (${networks[network].start})`
```
Comment (ChaituVR) [inline src/utils.ts:346]:
```suggestion
      `Snapshot (${snapshot}) must be 'latest' or greater than network start block (${networks[network].start})`
```

File: src/utils.ts [hunk 2]
```diff
@@ -530,6 +544,21 @@ export function getNumberWithOrdinal(n) {
...
+
+function isValidAddress(address: string) {
+  return isAddress(address) && address !== EMPTY_ADDRESS;
+}
+
+function isValidSnapshot(snapshot: number | string, network: string) {
+  return (
+    (typeof snapshot === 'number' && snapshot >= networks[network].start) ||
+    snapshot === 'latest'
+  );
```
Comment (ChaituVR) [inline src/utils.ts:559]:
```suggestion
  return (
    snapshot === 'latest' ||
    (typeof snapshot === 'number' && snapshot >= networks[network].start)
  );
```

**Review by ChaituVR — APPROVED** [review]
tAck

File: src/utils.spec.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+          'author is an invalid address',
+          { author: 'test-address' },
+          /invalid author/i
+        ],
+        ['network is not valid', { network: 'mainnet' }, /invalid network/i],
+        ['network is empty', { network: '' }, /invalid network/i],
+        [
+          'snapshot is smaller than start block',
+          { snapshot: 1234 },
+          /snapshot \([0-9]+\) must be greater than network start block/i
```
Comment (ChaituVR) [inline src/utils.spec.js:70]:
```suggestion
          /snapshot \([0-9]+\) must be 'latest' or greater than network start block/i
```

**Pull request merged**